cart.go

```diff
@@ -13,60 +13,131 @@ import (
-	JobId        string `json:"job_id"`
+	JobID        string `json:"job_id"`
 	WorkflowName string `json:"workflow_name"`
-	WorkflowId   string `json:"workflow_id"`
+	WorkflowID   string `json:"workflow_id"`
 }
 
 type build struct {
 	BuildNum  int       `json:"build_num"`
 	Revision  string    `json:"vcs_revision"`
 	Workflows *workflow `json:"workflows"` // plural name but singleton struct
+
+	// We want to skip bad builds, and perhaps print the others so that if
+	// there's a mismatch from expectations, folks might notice.
+	Status   string `json:"status"`
+	Subject  string `json:"subject"`
+	StopTime string `json:"stop_time"`
 }
 
 type artifact struct {
 	URL string `json:"url"`
 }
 
+// FilterSet is the collection of attributes upon which we filter the results
+// from Circle CI (or provide in URL to pre-filter).
+type FilterSet struct {
+	branch    string
+	workflow  string
+	jobname   string
+	anyFlowID bool
+}
+
+// Expander is used to take strings containing ${var} and interpolate them,
+// so that we don't have URLs which have %s/%s/%s and cross-referencing across
+// places to figure out which those fields are.
+type Expander map[string]string
+
+// Get is just a map lookup which panics, as a function for use with os.Expand
+func (e Expander) Get(key string) string {
+	if val, ok := e[key]; ok {
+		return val
+	}
+	// There is no recovery, we don't want to pass a bad URL out, we're
+	// a client tool and we'll need to fix the hardcoded template strings.
+	panic("bad key " + key)
+}
+
+// Expand converts "${foo}/${bar}" into "football/goal".
+// It also handles some $foo without parens, but we avoid using that.
+func (e *Expander) Expand(src string) string {
+	return os.Expand(src, e.Get)
+}
+
 var (
-	project               string
-	branch                string
-	buildNum              int
-	circleToken           string
-	outputPath            string
-	workflowArtifactBuild string
-	workflowDepth         int
-	verbose, dryRun       bool
+	circleToken     string
+	filter          FilterSet
+	verbose, dryRun bool
 )
 
 func main() {
+	var (
+		project             string
+		buildNum            int
+		outputPath          string
+		retrieveBuildsCount int
+	)
+
 	log.SetFlags(log.Lshortfile)
 	log.SetOutput(os.Stderr)
 
-	flag.StringVar(&project, "repo", "", "github `username/repo`")
-	flag.StringVar(&branch, "branch", "master", "search builds for branch `name`")
-	flag.IntVar(&buildNum, "build", 0, "get artifact for build number, ignoring branch")
-	flag.StringVar(&workflowArtifactBuild, "workflow-artifact-build", "", "if using a workflow, look for artifacts under build: `name`")
-	flag.IntVar(&workflowDepth, "workflow-depth", defaultWorkflowDepth, "how many workflow builds to check")
 	flag.StringVar(&circleToken, "token", "", "CircleCI auth token")
 	flag.StringVar(&outputPath, "o", "", "output file `path`")
 	flag.BoolVar(&verbose, "v", false, "verbose output")
 	flag.BoolVar(&dryRun, "n", false, "skip artifact download")
```

Comment:
While this line/arg didn’t change in this PR, this PR does change the command line interface for `cart`, so making a request:

- [ ] Change `-n` to `-dry-run` (or add an alias)